File: tests/integrated/test-laplace/test_laplace.cxx
/*
 * Laplacian inversion
 *
 */

#include <bout.hxx>
#include <invert_laplace.hxx>
#include <field_factory.hxx>

int main(int argc, char** argv) {

  // Initialise BOUT++, setting up mesh
  BoutInitialise(argc, argv);

  FieldFactory f{bout::globals::mesh};
  Options dump;

  Field3D input = f.create3D("(1-gauss(x-0.5,0.2))*gauss(y-pi)*gauss(z-pi)");
  Field2D a = f.create2D("gauss(x) * sin(y)");
  Field2D c = f.create2D("sin(x) * gauss(x-0.5) * gauss(y-pi)");
  Field2D d = f.create2D("y - pi/2");
  dump["input"] = input;
  dump["a"] = a;
  dump["c"] = c;
  dump["d"] = d;

  auto lap = std::unique_ptr<Laplacian>{Laplacian::create()};

  lap->setCoefA(0.0);
  lap->setCoefC(1.0);
  lap->setCoefD(1.0);
  dump["flag0"] = lap->solve(input);
  lap->setInnerBoundaryFlags(INVERT_DC_GRAD + INVERT_AC_GRAD);
  dump["flag3"] = lap->solve(input);

  lap->setCoefA(a);
  lap->setInnerBoundaryFlags(0);
  dump["flag0a"] = lap->solve(input);
  lap->setInnerBoundaryFlags(INVERT_DC_GRAD + INVERT_AC_GRAD);
  dump["flag3a"] = lap->solve(input);

  lap->setCoefC(c);
  lap->setInnerBoundaryFlags(0);
  dump["flag0ac"] = lap->solve(input);
  lap->setInnerBoundaryFlags(INVERT_DC_GRAD + INVERT_AC_GRAD);
  dump["flag3ac"] = lap->solve(input);

  lap->setCoefC(1.0);
  lap->setCoefD(d);
  lap->setInnerBoundaryFlags(0);
  dump["flag0ad"] = lap->solve(input);
  lap->setInnerBoundaryFlags(INVERT_DC_GRAD + INVERT_AC_GRAD);
  dump["flag3ad"] = lap->solve(input);

  /// Test new interface and INVERT_IN/OUT_SET flags

  Field2D set_to = f.create2D("cos(2*y)*(x - 0.5)");
  dump["set_to"] = set_to;

  lap->setCoefA(0.0);
  lap->setCoefC(1.0);
  lap->setCoefD(1.0);

  lap->setInnerBoundaryFlags(INVERT_SET);
  dump["flagis"] = lap->solve(input, set_to);
  lap->setInnerBoundaryFlags(0);
  lap->setOuterBoundaryFlags(INVERT_SET);
  dump["flagos"] = lap->solve(input, set_to);

  lap->setCoefA(a);
  lap->setInnerBoundaryFlags(INVERT_SET);
  lap->setOuterBoundaryFlags(0);
  lap->setOuterBoundaryFlags(0);
  dump["flagisa"] = lap->solve(input, set_to);
  lap->setInnerBoundaryFlags(0);
  lap->setOuterBoundaryFlags(INVERT_SET);
  dump["flagosa"] = lap->solve(input, set_to);

  lap->setCoefC(c);
  lap->setInnerBoundaryFlags(INVERT_SET);
  lap->setOuterBoundaryFlags(0);
  dump["flagisac"] = lap->solve(input, set_to);
  lap->setInnerBoundaryFlags(0);
  lap->setOuterBoundaryFlags(INVERT_SET);
  dump["flagosac"] = lap->solve(input, set_to);

  lap->setCoefC(1.0);
  lap->setCoefD(d);
  lap->setInnerBoundaryFlags(INVERT_SET);
  lap->setOuterBoundaryFlags(0);
  dump["flagisad"] = lap->solve(input, set_to);
  lap->setInnerBoundaryFlags(0);
  lap->setOuterBoundaryFlags(INVERT_SET);
  dump["flagosad"] = lap->solve(input, set_to);

  bout::writeDefaultOutputFile(dump);

  MPI_Barrier(BoutComm::get()); // Wait for all processors to write data

  bout::checkForUnusedOptions();
  BoutFinalise();
  return 0;
}

File: tests/integrated/test-laplace/test_laplace.py
#!/usr/bin/env python3
from pathlib import Path

#
# Run the test, compare results against the benchmark
#

# requires: not metric_3d
# Requires: netcdf
# Cores: 4

import pytest
import numpy.testing as npt
from boututils.run_wrapper import shell, launch_safe
from boutdata.collect import collect, create_cache


# Variables to compare
vars = [
    "flag0",
    "flag3",
    "flagis",
    "flagos",
    "flag0a",
    "flag3a",
    "flagisa",
    "flagosa",
    "flag0ac",
    "flag3ac",
    "flagisac",
    "flagosac",
    "flag0ad",
    "flag3ad",
    "flagisad",
    "flagosad",
]
tol = 1e-6  # Absolute tolerance


@pytest.fixture(scope="module")
def benchmark_data():
    """Fixture to load benchmark data once for the entire test session."""

    source_data_dir = str(Path(__file__).parent / "data")
    return {
        v: collect(v, path=source_data_dir, prefix="benchmark", info=False)
        for v in vars
    }


@pytest.mark.parametrize("solver", ["cyclic", "pcr", "pcr_thomas"])
@pytest.mark.parametrize("nproc", [1, 2, 4])
def test_laplace(solver, nproc, benchmark_data):

    nxpe = 2 if nproc > 2 else 1
    cmd = f"./test_laplace NXPE={nxpe} laplace:type={solver}"

    shell("rm data/BOUT.dmp.*.nc")

    s, out = launch_safe(cmd, nproc=nproc, mthread=1, pipe=True)

    with open(f"run.log.{nproc}", "w") as f:
        f.write(out)

    cache = create_cache(path="data", prefix="BOUT.dmp")

    # Collect output data
    for v in vars:
        result = collect(v, path="data", info=False, datafile_cache=cache)
        npt.assert_allclose(
            result,
            benchmark_data[v],
            atol=tol,
            rtol=tol,
            err_msg=f"Failed checking variable: {v}",
        )

    # Only check FieldPerps on one processor because reading them in is
    # quite annoying on multiple cores due to mismatched global y indices
    if nproc == 1:
        for v in ["flag0_perp", "flag3_perp"]:
            result = collect(v, path="data", info=False, datafile_cache=cache)

            # Compare benchmark and output
            npt.assert_allclose(
                result,
                benchmark_data[v.replace("_perp", "")][:, 0, :],
                atol=tol,
                rtol=tol,
                err_msg=f"Failed checking perpendicular variable: {v}",
            )
